### PR TITLE
gvim ligatures support with optimized screen output

### PR DESCRIPTION
**Pull request by dusanx**
PR: gvim ligatures support with optimized screen output

- In gvim, mixed content strings (ascii / multibyte utf-8) that are printed to screen, instead of falling down to slower Pango output are now split into ascii and utf-8 parts. That way we print as much as possible in a much more optimized way, trough pre-cached glyphs. This is speed and code optimization on it's own.

- Some programming fonts offer ligatures, combinations of plain ascii characters, that combined create different, usualy better looking shapes. This PR introduces `guiligatures` global option. Default value is empty string, ligatures are not shown, so this makes non-breaking change.

- User can `:set guiligatures` to an empty string or add as many characters as needed, depending on the font capabilities and computer speed. This is very flexible compared to other editor that offer ligatures support.

- PR should fully respect `:h coding-style` but I am ready to make amends if something slipped.

Example:
```
:set guiligatures=!\"#$%&()*+-./:<=>?@[]^_{\|~
```

Recap: Being speed optimization and non-breaking change, this PR should be strong candidate for merging.

Ready to answer any question. Thank you.

### CODE REVIEW

**Comment by dusanx** [comment]
Amending this one warning. one sec.

**Comment by chrisbra** [comment]
thank you, I am looking forward for that change.

**Comment by dusanx** [comment]
That's it, ready to merge. Thanks @chrisbra for your help to quickly find where is what.

**Comment by dusanx** [comment]
Hold on, I found edge case where this PR creates weird artifacts with utf-8 characters in statusline. Debugging.

**Comment by dusanx** [comment]
Fixed, sorry about that, it was last minute edge case catch...

**Comment by dusanx** [comment]
Just to confirm, nothing else to do here, ready to merge.

**Comment by dusanx** [comment]
Since it is my first vim PR, question: Are this failed checks something I should look at or usual to be ignored stuff? Looking at few of them looks like it is not me related?

**Comment by matveyt** [comment]
@dusanx `gui_set_ligatures()` must be inside `#if defined(FEAT_GUI_GTK)`. Currently it's not and for this reason Windows build fails.

Also, there's some problem with one of GUI tests (testdir/test_gui.vim::Test_encoding_conversion). Not sure why but it's SIGSEGV.

**Comment by dusanx** [comment]
@matveyt thanks, I'll take a look. Since I don't have windows and this is my first PR, maybe better that somebody jumps in and makes this accepted by windows? I can of course apply `#if defined(FEAT_GUI_GTK)` but only in a blind mode and I can't trigger CI tests myself so may be slow cycle to get it right.

Re (testdir/test_gui.vim::Test_encoding_conversion) I'll certainly try to see if I did something unexpected.

**Comment by dusanx** [comment]
set encoding=latin1... I'll need a day to test this. Thanks for your help, I'll commit soon.

**Comment by dusanx** [comment]
Updated, ready for tests:

1. `# ifdef FEAT_GUI_GTK` hopefully done right way, blind edit since I don't have windows to test.
2. Segfault completely my fault, with latin1 string goes trough ` if (output_conv.vc_type != CONV_NONE)`, I was releasing  `vim_free(conv_buf);` too early.

This two should be clear now. Let's see what tests say.

**Comment by matveyt** [comment]
Code in `optionstr.c` also must be guarded. I've created a PR into your repo.

**Comment by dusanx** [comment]
@matveyt thanks for helping me, AppVeyor has finally passed. Of course next PR will require less help, still learning where is what in vim.

**Comment by matveyt** [comment]
@dusanx Don't mention it. We still have to wait for Bram to review & accept this.

**Comment by dusanx** [comment]
Hmhm, I don't like this one failing test but I'll take a look immediately.

Btw when this PR is merged someone can teach me if and how I can run this tests locally, before sending PR?

**Comment by dusanx** [comment]
Got it, sorry about this, I'll push in 10-20 minutes...

**Comment by dusanx** [comment]
Ok, so:
1. I am really starting to love this tests
2. This last bugfix was really embarrassing... please don't look anybody. I _maybe_ had to expand buffer to 256 chars in order to test whole range of char_u ascii but that is most probably just a rumour.

**Comment by dusanx** [comment]
How green it is? Super green :smile:

**Comment by brammool** [comment]
Please!  No Chris Tucker imitations!  :-)

**Comment by dusanx** [comment]
Haha good catch :)

**Comment by dusanx** [comment]
Short question since I am new here: my master is now five commits behind. Since I don't think there is anything else to do here and only waits for review, should I pull master, rebase, push this branch again or just wait? What is usually expected for PR?

**Comment by vim-ml** [comment]
Hi,

On Sat, Oct 2, 2021 at 2:50 PM dusanx ***@***.***> wrote:

> Short question since I am new here: my master is now five commits behind.
> Since I don't think there is anything else to do here and only waits for
> review, should I pull master, rebase, push this branch again or just wait?
> What is usually expected for PR?
>
>
>
There is no need to update your PR to the latest. When Bram decides to
merge your
changes, he will review the changes and then commit them.

I usually update my PRs once in a while to the latest to avoid conflicts
with the latest
changes so that when Bram merges the changes it goes smoothly.

Regards,
Yegappan

**Comment by brammool** [comment]
So long as the pull request says "This branch has no conflicts with the base branch" there is no need to update.

**Comment by brammool** [comment]
A few remarks:
- Instead of using vim_memset() you can use CLEAR_FIELD().  Then the size is automatic.
- We have the style rule that below declarations there is an empty line.
-  declarations must go at the start of a block, not after a statement.  Not all compilers handle the C99 style (MSVC for example, even though this code won't be used on windows it is still good to keep this rule)
- in gui_set_ligatures() you don't need the "ch" variable, just use *p directly.
- the comment above gui_set_ligatures() mentions 128, the array is now 256 long. The top 128 are always zero, but I think that's a nice way to stay on the safe side, so long as the index is an unsigned char.
- in gui_set_ligatures() when a non-ASCII character is encountered this should result in an error, and the old option value should be kept.
- We need a test for the option.  You can't check the effect, but at least check some valid values and some invalid values (with the error just added).
-  the declaration of gui_gtk2_draw_string_ext() is more than 80 characters, in that case we use one argument per line.

**Comment by dusanx** [comment]
Thanks for the review. I'll push update in day or two, everything is easy to follow. I was also on the edge of using smaller 96 char map (since we are only interested in ascii 32-127) but that requires few more if statements in a tight loop, so as you say 256 keeps us on the safe side.

> should result in an error,

How do we assign new error numbers?

> We need a test for the option. You can't check the effect, but at least check some valid values and some invalid values (with the error just added).

This is the only thing I don't know (yet) how to implement and not sure how to test test locally. Is there any doc/guide how to add test? Or in simpler terms what proper existing test file where I should add this one?

Thanks!

**Comment by dusanx** [comment]
Yes! I have single-handedly vanquished the beast! 

- All variable declarations where they should be.
- E1240 was empty so I took it, if there is better method than picking first empty Exxx value please change or let me know.
- Testing, asserting, throwing errors if needed.

Assuming that E1240 can be used this is ready to merge.

**Comment by dusanx** [comment]
Can I ask somebody to take a look at

https://github.com/dusanx/vim/blob/64625ea3fe245aa2b426a29eb97f55b9f4d1453b/src/testdir/test_gui.vim#L569-L588

and tell me why Test_set_guiligatures() fails on one windows and one mac test? Do I need to wrap all of that into

```
if !g:x11_based_gui
  let skipped = g:not_supported . 'guiligatures'
else
```
then test for gtk? I don't have windows or mac so I can only blind edit, push, wait for hours too see results -- not really efficient debug cycle.

**Comment by dusanx** [comment]
> The test looks good, as far as I can tell. Do you have a failing logfile somewhere?

Sadly no, I just pushed another try with if !g:x11_based_gui and it was a bad idea -- I can see tests falling all over. Let me wait for failing tests to finish, revert change to previous one, push, and I'll @ you for help. I am not sure I can get tests right without help on first PR.

**Comment by dusanx** [comment]
Guys, something is wrong with CI tests again -- please check 'failures' above. I don't think there is anything I can do on my side.

**Comment by dusanx** [comment]
If somebody can flip switch to run this one failed test I would appreciate the gesture, I can't do it myself and pretty sure it was/is CI problem. Alternatively if it is something I caused I would be thankful for an explanation how not to do it again. Anyway I have some spare time and few more improvements I would like to work on but they are based on this one so waiting to see how first one ends. Thank you.

**Comment by chrisbra** [comment]
I cannot run just one single failing ci instance, I can only re-run all the CI systems and possibly we have another flaky test failing. 
So I think you are fine currently.
